File: 463/island-perimeter.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
function islandPerimeter(grid) {
    if (grid.length == 0) {
        return 0;
    }

    let [m, n] = [grid.length, grid[0].length];
    let peri = 0;

    const dir = [[1, 0], [-1, 0], [0, 1], [0, -1]];

    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            if (grid[i][j] == 0)
                continue;

            for (let l = 0; l < 4; l++) {
                let neighbourX = i + dir[l][0];
                let neighbourY = j + dir[l][1];

                if (neighbourX < 0 || neighbourX >= m || neighbourY < 0 || neighbourY >= n) {
                    peri++;
                    continue;
                }

                peri += !grid[neighbourX][neighbourY];
            }
        }
    }
    return peri;
}
/**
 * @param {number[][]} grid
 * @return {number}
 */
function islandPerimeter(grid) {
    if (grid.length == 0) {
        return 0;
    }

    let [m, n] = [grid.length, grid[0].length];
    let peri = 0;

    const dir = [[1, 0], [-1, 0], [0, 1], [0, -1]];

    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            if (grid[i][j] == 0)
                continue;

            for (let l = 0; l < 4; l++) {
                let neighbourX = i + dir[l][0];
                let neighbourY = j + dir[l][1];

                if (neighbourX < 0 || neighbourX >= m || neighbourY < 0 || neighbourY >= n) {
                    peri++;
                    continue;
                }

                peri += !grid[neighbourX][neighbourY];
            }
        }
    }
    return peri;
}

File: 463/island-perimeter.py
class Solution:
    def islandPerimeter(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        m, n = len(grid), len(grid[0])
        if m == 0:
            return 0

        peri = 0
        dirs = [[1, 0], [-1, 0], [0, 1], [0, -1]]
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 0:
                    continue
                for dir in dirs:
                    neighbour_x = i + dir[0]
                    neighbour_y = j + dir[1]
                    if (neighbour_x < 0 or neighbour_x >= m) or (neighbour_y < 0 or neighbour_y >= n):
                        peri += 1
                        continue
                    peri += 0 if grid[neighbour_x][neighbour_y] else 1

        return peri
